src/Workspaces/Core/Portable/Workspace/Host/DocumentService/IDocumentServiceProvider.cs

```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    internal interface IDocumentServiceProvider
+    {
+        /// <summary>
+        /// Gets a document specific service provided by the host identified by the service type. 
+        /// If the host does not provide the service, this method returns null.
+        /// </summary>
+        TService GetService<TService>() where TService : class, IDocumentService;
```

Comment:
see below comment.  it feels like the use cases we have identified would work with no-op cases.  allowed features that we want to use this stuff to not need fallback cases for how to deal with null.  This feels more appropriate to me.  Take excerpts for example, it feels like we shoudl always be able to get the excerpt (moving the code that is in FAR over to a common location), but that razor could override the excerpt behavior.  then FAR remains blissfully ignorant of how this works, but can just trust that it can get excerpts.  This also means any future features have hte same capability, instead of them having to agree on a common codepath for the fallback case (or worse, not having a common fallback).

Reply:
we will do that once we are settled down on this design. we are not even sure this will be the pattern we will settle at the end. those can be done in separate PRs.